### ISSUE TITLE
[Hilt] Getting NullPointerException when mocking the view model in UI tests

### ISSUE THREAD

**Issue by oanamariabalaita** [google/dagger#3300]
Kotlin 1.6.10
Dagger/Hilt: 2.41
Gradle: 7.4.1
AGP: 7.1.2

```kotlin
@HiltAndroidTest
class OnboardingTest : BaseUiTest() {

    @BindValue
    @JvmField
    val viewModel: OnboardingViewModel = mock()
    
}
```
When compiling I get the following:

```
java.lang.NullPointerException: Null reference used for synchronization (monitor-enter)
at androidx.lifecycle.ViewModel.setTagIfAbsent(ViewModel.java:156)
at androidx.lifecycle.AbstractSavedStateViewModelFactory.create(AbstractSavedStateViewModelFactory.java:70)
at androidx.lifecycle.AbstractSavedStateViewModelFactory.create(AbstractSavedStateViewModelFactory.java:84)
at dagger.hilt.android.internal.lifecycle.HiltViewModelFactory.create(HiltViewModelFactory.java:109)
at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.java:187)
at androidx.lifecycle.ViewModelProvider.get(ViewModelProvider.java:150)
at androidx.lifecycle.ViewModelLazy.getValue(ViewModelProvider.kt:54)
at androidx.lifecycle.ViewModelLazy.getValue(ViewModelProvider.kt:41)
at OnboardingFragment.getViewModel(OnboardingFragment.kt:51)
at OnboardingFragment.initViewBinding(OnboardingFragment.kt:72)
at BaseFragment.onCreateView(BaseFragment.kt:18)
at androidx.fragment.app.Fragment.performCreateView(Fragment.java:2995)
at androidx.fragment.app.FragmentStateManager.createView(FragmentStateManager.java:523)
at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:261)
at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:1840)
at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:1758)
at androidx.fragment.app.FragmentManager.execPendingActions(FragmentManager.java:1701)
at androidx.fragment.app.FragmentManager$4.run(FragmentManager.java:488)
at android.os.Handler.handleCallback(Handler.java:938)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loopOnce(Looper.java:201)
at android.os.Looper.loop(Looper.java:288)
at android.app.ActivityThread.main(ActivityThread.java:7839)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
```

**Comment by Chang-Eric** [comment]
I don't think this is Hilt related, this probably has to do with just the mock returning null in places where it is unexpected. In general, you shouldn't be mocking a ViewModel in the same way you shouldn't mock a Fragment or an Activity. These are tied too much into Android in complicated ways that you shouldn't expect to be able to just replace it with a mock. Instead, you should use the real definition of the ViewModel and replace the Dagger bindings it uses as needed instead. Alternatively, you can completely hide the usage of the ViewModel behind an interface and replace that.

There's also some information about the testing philosophy Hilt uses here https://dagger.dev/hilt/testing-philosophy that can help explain at a high level the kind of testing we are aiming to support.

**Comment by oanamariabalaita** [comment]
Closing this as the issue is fixed in version [2.44](https://github.com/google/dagger/releases/tag/dagger-2.44.).
